Add application with LTI Provider

idegtiarov opened the pull request:
PR adds application with LTI Provider to the StarCellBio project

Documentation describing configuration on Provider and Consumer sides could be found in the document by the link: [LTI Configurations](https://docs.google.com/document/d/1vW0HEbjVvxdG6ezZM-mkuijiJMqemG-745IXSHtGdb0/edit?usp=sharing)

idegtiarov committed:
Create lti_provider application with the base provider's logic

idegtiarov committed:
Add fixes and changes in the ILT provider

idegtiarov committed:
Remove useless models with required changes

idegtiarov committed:
Move validation and checking functionality to the validator.

Add doc strings.
Provide partial refactoring.

idegtiarov committed:
Add login process of the newly created user

idegtiarov committed:
Fix issue with user login process

idegtiarov committed:
Add adjustment for the added resources via LTI

lti/launch/course/<course_id> -- specifying the course
lti/launch/course/<course_id>/assignment/<assignment_id> -- specifying
certain assignment in the course
lti/launch/course/<course_id>/assignment/<assignment_id>/\
experiment/<experiment_id> -- specifying experiment in the assignment

idegtiarov committed:
Add validator unit tests

idegtiarov committed:
Add models and views unit tests.

Add small changes in models and views files.

idegtiarov committed:
Improve URL pattern and redirection to the experiment launch

idegtiarov committed:
Fix launch_url in config view

idegtiarov committed:
Pin mock and improve six packages

idegtiarov committed:
Improve requests package versions in requirements

Solving Travis tests issues

idegtiarov committed:
Fix view test and change lti_settings to the default state

idegtiarov committed:
Fix settings, and tests to improve Travic CI checking

annagav commented on StarCellBio/settings.py:
```diff
@@ -112,7 +112,6 @@
 MIDDLEWARE_CLASSES = (
     'django.middleware.common.CommonMiddleware',
     'django.contrib.sessions.middleware.SessionMiddleware',
-    'django.middleware.csrf.CsrfViewMiddleware',
```
Why are we removing this?

idegtiarov replied:
I've added this middleware in commit #31106d6 and found that it broke frontend tests that is why I removed it as it was in initial state.

blarghmatey committed:
Updated log config to work in deployment environment

The updated log configuration was causing startup errors due to the
fact that was trying to write files and create directories where it
doesn't have permission to do so on the host.

auraz commented on lti_provider/lti_settings.py:
```diff
@@ -0,0 +1,8 @@
+# Switch on debug mode for LTI requests
+DEBUG_LTI = getattr(settings, 'DEBUG_LTI', True)
+
+# Uncoment to switch off LTI ssl requirements (only for development mode)
```
UncomMent, typo.

idegtiarov replied:
done

auraz commented on lti_provider/models.py:
```diff
@@ -0,0 +1,69 @@
+        """
+        if self.scb_user:
+            self.scb_user.backend = 'allauth.account.auth_backends.AuthenticationBackend'
+            logger.debug("Start User {} logining...".format(self.scb_user.username))
```
Typo, logging.

idegtiarov replied:
fixed

auraz commented on lti_provider/models.py:
```diff
@@ -0,0 +1,69 @@
+            self.scb_user.backend = 'allauth.account.auth_backends.AuthenticationBackend'
+            logger.debug("Start User {} logining...".format(self.scb_user.username))
+            login(request, self.scb_user)
+            logger.debug("Check User is authenticated:: {}".format(request.user.is_authenticated()))
```
Rephrase plz.

auraz commented on lti_provider/unit_tests/test_views.py:
```diff
@@ -0,0 +1,92 @@
+
+
+def mock_tool_provider():
+    m = Mock()
```
single value names are restricted by pep8, i suppose.

idegtiarov replied:
Not allowed to use only: l, O, and I variables.

auraz commented on lti_provider/validator.py:
```diff
@@ -0,0 +1,77 @@
+        consumer_expired_date = self.consumer.expiration_date
+        if consumer_expired_date and consumer_expired_date < today:
+            raise oauth1.OAuth1Error('Consumer Key is expired.')
+        return True
```
Where is noonce validated here?

idegtiarov replied:
We inherit oauth1.RequestValidator functions and timestamp is validated with `validate_timestamp_and_nonce` function. Nonce is not validating here.

auraz replied:
why it is passed and mentioned in name?

idegtiarov replied:
Because in Oauth1 authentication this two params used and validated in pairs.

auraz commented on lti_provider/validator.py:
```diff
@@ -0,0 +1,77 @@
+        :param request: LTI request
+        :return: boolean flag
+        """
+        logger.debug('Client key is validating...')
```
Please rephrase to be stateful.

idegtiarov replied:
done

auraz commented on lti_provider/views.py:
```diff
@@ -0,0 +1,114 @@
+    ]))
+
+    # maybe you've got some extensions
+    # FIXME(idegtiarov) there is no extentions right now, if they don't appear remove example.
```
Please clarify this during pull request review, or change comment text.

idegtiarov replied:
We are not interested in any extension parameters, I remove this comment

auraz commented on lti_provider/views.py:
```diff
@@ -0,0 +1,114 @@
+    if settings.DEBUG_LTI:
+        logger.debug("LTI request is {}valid".format('' if ok else 'not '))
+    if not ok:
+        # TODO(idegtiarov) add lti_error page if it is needed
```
Clean or fix TODO plz.

idegtiarov replied:
Done

auraz commented on lti_provider/views.py:
```diff
@@ -0,0 +1,114 @@
+
+    user_id = request_post.get('user_id')
+    if not user_id:
+        # TODO(idegtiarov) add lti_error page if it is needed
```
Clean or fix TODO plz.

idegtiarov replied:
done

auraz commented on lti_provider/views.py:
```diff
@@ -0,0 +1,114 @@
+    user.login(request)
+    url = reverse('home')
+    if not course_id or not Course.objects.filter(code=course_id).exists():
+        # TODO(idegtiarov) add lti_error page if it is needed
```
Clean or fix TODO plz.

idegtiarov replied:
done

auraz commented on lti_provider/views.py:
```diff
@@ -0,0 +1,114 @@
+        raise Http404('Course with the code {}, does not exist.'.format(course_id))
+    if assignment:
+        if not Assignment.objects.filter(assignmentID=assignment, courseID__code=course_id).exists():
+            # TODO(idegtiarov) add lti_error page if it is needed
```
Clean or fix TODO plz.

idegtiarov replied:
done

auraz commented on lti_provider/views.py:
```diff
@@ -0,0 +1,114 @@
+        tool_provider = DjangoToolProvider.from_django_request(request=request)
+        validator = RequestValidator()
+        ok = tool_provider.is_valid_request(validator)
+    except oauth1.OAuth1Error as err:
```
Can this error be raised here too DjangoToolProvider.from_django_request(request=request) ?

idegtiarov replied:
Here can be raised two other errors InvalidLTIRequestError and ValueError. Add them in except checking block.

auraz commented on lti_provider/views.py:
```diff
@@ -0,0 +1,114 @@
+    if not user.is_scb_user:
+        # NOTE(idegtiarov) connect user with the SCB user account
+        user.lti_to_scb_user(roles, course_id)
+        logger.debug('Check user was created {}'.format(user.is_scb_user))
```
please rephrase, what do you mean here?

idegtiarov replied:
done

idegtiarov committed:
Fix typos and comments form review.

idegtiarov committed:
Fix comments from review.

annagav commented:
@idegtiarov, While deploying to our qa site there were at least a couple of dependencies missing in requirements.txt. Could you please update it?

idegtiarov commented:
@annagav Yes sure I will check.

idegtiarov committed:
Merge branch 'develop' into ilt_provider

# Conflicts:
#	requirements.txt

idegtiarov commented:
@annagav I merge latest develop branch and run SCB in the new vagrant instance, requirements.txt covers all dependencies. If you know what packages should be added or updated please let me know.

annagav commented:
Functionality works great! 👍
